index.bs

```diff
@@ -369,6 +448,110 @@ Report long tasks {#report-long-tasks}
+
+    The <dfn>relevant frame timing info</dfn> for a {{Document}} |document| is its [=nearest same-origin root=]'s [=current frame timing info=].
+</div>
+
+<div algorithm="Report task start time">
+    To <dfn export>report task start time</dfn> given an {{DOMHighResTimeStamp}} |taskStartTime|, and a {{Document}} |document|:
+
+    1. Let |root| be |document|'s [=nearest same-origin root=].
+
+    1. If |root|'s [=current frame timing info=] is null,
+        then set |root|'s [=current frame timing info=] to a new [=frame timing info=] whose [=frame timing info/start time=] is |taskStartTime|.
+
+    1. Set |timingInfo|'s [=current frame timing info=]'s [=frame timing info/current task start time=] to |taskStartTime|.
+
+    1. If |timingInfo|'s [=frame timing info/start time=] is 0, then set |timingInfo|'s [=frame timing info/start time=] to |taskStartTime|.
+</div>
+
+<div algorithm="Report task end time">
+    To <dfn export>report task end time</dfn> given an {{DOMHighResTimeStamp}} |taskEndTime|, and a {{Document}} |document|:
+
+    1. Let |timingInfo| be |document|'s [=relevant frame timing info=].
+
+    1. If |timingInfo| is null, then return.
+
+        Note: This can occur if the browser becomes hidden during the sequence.
+
+    1. Let |safeTaskEndTime| be the [=relative high resolution time=] given |taskEndTime| and |document|'s [=relevant global object=].
+
+    1. Let |safeTaskStartTime| be the [=relative high resolution time=] given |timingInfo|'s [=frame timing info/current task start time=] and |document|'s [=relevant global object=].
+
+    1. Let |currentTaskDuration| be the [=duration=] between |safeTaskStartTime| and |safeTaskEndTime|.
+
+    1. If |currentTaskDuration| is greater than |timingInfo|'s [=frame timing info/longest task duration=], then set |timingInfo|'s [=frame timing info/longest task duration=] to |currentTaskDuration|.
+
+    1. If the user agent believes that updating the rendering of |document|'s [=node navigable=] would have no visible effect, then [=flush frame timing=] given |document| and return.
+</div>
+
+<div algorithm="Report rendering time">
+    To <dfn export>report rendering time</dfn> given a {{Document}} |document|, and a {{DOMHighResTimeStamp}} |styleAndLayoutStart|:
+
+    1. Let |timingInfo| be |document|'s [=relevant frame timing info=].
+
+    1. If |timingInfo| is null, then return.
+
+        Note: This can occur if the browser becomes hidden during the sequence.
+
+    1. Set |timingInfo|'s [=frame timing info/update the rendering start time=] to |timingInfo|'s [=frame timing info/current task start time=].
+
+    1. Set |timingInfo|'s [=frame timing info/style and layout start time=] to |styleAndLayoutStart|.
+
+    1. [=Flush frame timing=] given |document| and the [=unsafe shared current time=].
+</div>
+
+<div algorithm="Queue frame timing">
+    To <dfn export>flush frame timing</dfn> given a {{Document}} |document| and a {{DOMHighResTimeStamp}} |endTime|:
```

Comment:
Generally, it might be better to prefix all the unsafe timestamps with "unsafe" (or something similar), rather than prefixing the safe ones.